<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Project: ‘BRFC’ - Behavior risk factors &amp; cancer project.</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="report_files/libs/clipboard/clipboard.min.js"></script>
<script src="report_files/libs/quarto-html/quarto.js"></script>
<script src="report_files/libs/quarto-html/popper.min.js"></script>
<script src="report_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="report_files/libs/quarto-html/anchor.min.js"></script>
<link href="report_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="report_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="report_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="report_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="report_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Project: ‘BRFC’ - Behavior risk factors &amp; cancer project.</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<div class="cell" data-execution_count="80">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> sqlalchemy</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.model_selection <span class="im">import</span> train_test_split</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.linear_model <span class="im">import</span> LogisticRegression</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.metrics <span class="im">import</span> classification_report, roc_curve, auc, roc_auc_score</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> warnings</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>warnings.simplefilter(action<span class="op">=</span><span class="st">'ignore'</span>, category<span class="op">=</span><span class="pp">FutureWarning</span>)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="co">#use provided anaconda environment file: made_project_environment.yml to setup an environment for this file</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<section id="motivation" class="level3">
<h3 class="anchored" data-anchor-id="motivation">Motivation</h3>
<p>Cancer is one of the leading number of deaths, bronchus and lung cancer deaths are now ranked 6th among the leading causes of deaths (https://www.who.int/news-room/fact-sheets/detail/the-top-10-causes-of-death). Also obesity is a world wide health issue and influences diseases. Rising numbers in cancer and obesity are alerting and therefore it is of high interest to reveal any relations of obesity/behavioral risk factors and cancer in order to apply countermeasures.</p>
<p>This project aims to uncover relations of behavioral risk factors and cancer.</p>
<p>The hypothesis for this project: Behavioural risk factors are influencing the chance to develop cancer. and Behavioural risk factors are influencing the chance of mutation(s) in genes which are related to obesity.</p>
</section>
<section id="what-are-behavioural-risk-factors" class="level3">
<h3 class="anchored" data-anchor-id="what-are-behavioural-risk-factors">What are behavioural risk factors?</h3>
<p>Behavioural risk factors have a strong influence on health and can be:<br>
- Tocacco use<br>
- obesity<br>
- alcohol consumption<br>
and many more.</p>
<p>For this project the results of a nation wide study in the United States, the ‘BRFSS Behavioral Risk Factor Surveillance System’ is used. It is run by the Centers for Disease Control and Prevention. Citizens are regulary asked questions in order to collect data for the following statements:<br>
- Percent of adults aged 18 years and older who have an overweight classification<br>
- Percent of adults aged 18 years and older who have obesity<br>
- Percent of adults who engage in muscle-strengthening activities on 2 or more days a week<br>
- Percent of adults who engage in no leisure-time physical activity<br>
and more.<br>
<a href="https://www.cdc.gov/brfss/about/brfss_faq.htm">BRFSS FAQ</a></p>
</section>
<section id="what-is-gdc" class="level3">
<h3 class="anchored" data-anchor-id="what-is-gdc">What is GDC?</h3>
<p>This project includes cancer data from the Genomic Data Commons (GDC). This is a reseach program of the National Cancer Institute (NCI) in the United States. GDCs mission is to provide a unified repository to the cancer research community to support research to help all people to live longer, healthier lives. <a href="https://gdc.cancer.gov/about-gdc">GDC - About the GDC</a></p>
</section>
</section>
<section id="methods" class="level2">
<h2 class="anchored" data-anchor-id="methods">Methods</h2>
<section id="datasources" class="level3">
<h3 class="anchored" data-anchor-id="datasources">Datasources</h3>
<section id="datasource1-graphql-api-gdc---nci-genomic-data-commons" class="level4">
<h4 class="anchored" data-anchor-id="datasource1-graphql-api-gdc---nci-genomic-data-commons">Datasource1 GraphQL API: GDC - NCI Genomic Data Commons</h4>
<p>License: Data usage agreement: https://gdc.cancer.gov/about-data/data-analysis-policies<br>
Data Model: https://gdc.cancer.gov/developers/gdc-data-model</p>
<p>Cancer and genomic data GraphQL API<br>
- Metadata URL: https://docs.gdc.cancer.gov/API/Users_Guide/GraphQL_Examples/<br>
- Data URL: https://api.gdc.cancer.gov/v0/graphql<br>
- Query: see query in pipeline code (data/gdc1a.py) starting from line 27: <a href="data/gdc1a.py#L27">Projects GraphQL query</a><br>
- Data Type: json</p>
</section>
<section id="datasource2-behavioral-risk-factor-surveillance" class="level4">
<h4 class="anchored" data-anchor-id="datasource2-behavioral-risk-factor-surveillance">Datasource2 Behavioral Risk Factor Surveillance</h4>
<p>License: https://opendefinition.org/licenses/odc-odbl/</p>
<p>Nutrition, Physical Activity, and Obesity - Behavioral Risk Factor Surveillance System<br>
- Metadata URL: https://catalog.data.gov/harvest/object/721fe106-9250-45d7-9093-1edacb565cd4<br>
- Data URL: https://catalog.data.gov/dataset/nutrition-physical-activity-and-obesity-behavioral-risk-factor-surveillance-system<br>
- Data Type: json/xml/rdf/csv</p>
<p>Centers for Disease Control and Prevention (CDC). Behavioral Risk Factor Surveillance System Survey Data. Atlanta, Georgia: U.S. Department of Health and Human Services, Centers for Disease Control and Prevention, [2023].</p>
</section>
</section>
<section id="data-pipeline" class="level3">
<h3 class="anchored" data-anchor-id="data-pipeline">Data pipeline</h3>
<section id="datasource1-graphql-api-gdc" class="level4">
<h4 class="anchored" data-anchor-id="datasource1-graphql-api-gdc">Datasource1 GraphQL API: GDC</h4>
<p>File: <a href="data/gdc1a.py">GDC pipeline</a><br>
Python pipeline to download via GraphQL and store the data into objects.<br>
One aspect is that GraphQL is a WebAPI which is not downloading the full dataset at one time. A cursor-Based pagination approach is used to get further data until no data is left. This is implemented in the python pipeline.</p>
</section>
<section id="datasource2-brfss" class="level4">
<h4 class="anchored" data-anchor-id="datasource2-brfss">Datasource2 BRFSS</h4>
<p>File: <a href="data/brfss.jv">BRFSS pipeline</a><br>
For this pipeline the domain-specific language and runtime ‘Jayvee’ is used. It is very easy and straightforward, therefore it was chosen for this pipeline in order to download and store the csv file.<br>
<a href="https://jvalue.github.io/jayvee/">Jayvee Project</a></p>
<p>Both of the pipelines are storing the data into a sqlite database.</p>
</section>
<section id="transformation" class="level4">
<h4 class="anchored" data-anchor-id="transformation">Transformation</h4>
<section id="genes" class="level5">
<h5 class="anchored" data-anchor-id="genes">Genes</h5>
<p>Specific genes are chosen to be of interest in this project. Evaluation on genes contributing to obesity are shown in paper ‘<em>Genes contributing to obesity</em>’ <a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>.<br>
The following genes were checked in the download of the GDC data and if available (=mutated) a variable is set to True in the sink (sqlite) database.<br>
- SIM1<br>
- POMC<br>
- LEPR<br>
- MRAP2<br>
- ADCY3<br>
- NTRK2<br>
- MC4R<br>
- KSR2<br>
- LEP<br>
- PCSK1<br>
- BDNF<br>
- SH2B1</p>
</section>
<section id="age-in-days" class="level5">
<h5 class="anchored" data-anchor-id="age-in-days">Age in days</h5>
<p>GDC data contains the field ‘age_at_diagnosis’ which represents the age in days. Before saving to the sqlite database it is additionally transformed to a field representing the age in years using the conversion factor 365.25 defined by GDC (<a href="https://docs.gdc.cancer.gov/Data_Portal/Users_Guide/Advanced_Search/#age-at-diagnosis-unit-in-days">Unit in days</a>)</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (aad):</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>        aad <span class="op">=</span> <span class="bu">round</span>(aad<span class="op">/</span><span class="fl">365.25</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>        diagnosis_dict[<span class="st">'age_at_diagnosis_years'</span>] <span class="op">=</span> aad</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
</section>
<section id="limitations" class="level3">
<h3 class="anchored" data-anchor-id="limitations">Limitations</h3>
<section id="data-availability-in-gdc" class="level4">
<h4 class="anchored" data-anchor-id="data-availability-in-gdc">Data availability in GDC</h4>
<p>The data in GDC shows that there are gaps and the cancer data is definitly not complete. This is clear since submitters are not forced to provide the data and only a subset of data is stored at GDC and open to the public.</p>
<div class="cell" data-execution_count="81">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>dbEngine<span class="op">=</span>sqlalchemy.create_engine(<span class="st">'sqlite:///../data/gdc.sqlite'</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> dbEngine.<span class="ex">connect</span>() <span class="im">as</span> conn:</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    sqlstatement_obese <span class="op">=</span> <span class="st">"select year_of_diagnosis, count(*) as c from diagnoses where year_of_diagnosis is not null group by year_of_diagnosis"</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    gdc <span class="op">=</span> pd.read_sql_query(sqlstatement_obese, conn)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    gdc[<span class="st">'year_of_diagnosis'</span>] <span class="op">=</span> gdc[<span class="st">'year_of_diagnosis'</span>].astype(<span class="bu">int</span>)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>    gdc <span class="op">=</span> gdc.set_index(<span class="st">'year_of_diagnosis'</span>)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> gdc[<span class="st">'c'</span>].plot(kind<span class="op">=</span><span class="st">'bar'</span>)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>ax.<span class="bu">set</span>(xlabel<span class="op">=</span><span class="st">'Year'</span>, ylabel<span class="op">=</span><span class="st">'Number of cancer diagnoses'</span>, title<span class="op">=</span><span class="st">'Cancer diagnoses within GDC database'</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="report_files/figure-html/cell-3-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="us-data-in-gdc-and-age-at-diagnosis" class="level4">
<h4 class="anchored" data-anchor-id="us-data-in-gdc-and-age-at-diagnosis">US data in GDC and age at diagnosis</h4>
<p>The data in GDC where the country of residence at enrollment of the patient is United States has implications with the age at diagnosis data availability. The range of age for cancer diagnoses in US are<br>
- max age 29<br>
- number aged 18 and older within the gdc data for US = 3</p>
<p>However the BRFSS data holds data for persons aged 18 and older.</p>
<p>Therefore for this report GDC data is not filtered by ‘United States’ and opened to all or no definition on country of residence at enrollment with the constraint:<br>
- the evaluation may not be valid based on the BRFSS data relating to US only<br>
- Assumption: The inclusion of datasets from different countries will not change the evaluation since the behavioural risk factors can be generalized</p>
</section>
<section id="data-merge-brfss-and-gdc" class="level4">
<h4 class="anchored" data-anchor-id="data-merge-brfss-and-gdc">Data merge BRFSS and GDC</h4>
<p>BRFSS data is available from 2011 to 2022. In the graph above you can see that the data submission for GDC drastically drops in 2014. Since there is enough data available to support the report, we restrict the evaluation on the year 2011 which is available in both the BRFSS and GDC data source.</p>
</section>
</section>
<section id="correlation" class="level3">
<h3 class="anchored" data-anchor-id="correlation">Correlation</h3>
<p>The evaluation correlates the data present in year 2011 for all diagnoses and the BRFSS data for the statements:<br>
- overweight_prob: Percent of adults aged 18 years and older who have an overweight classification<br>
- obesity_prob: Percent of adults aged 18 years and older who have obesity<br>
- musclestr2ormore_prob: Percent of adults who engage in muscle-strengthening activities on 2 or more days a week<br>
- nophyac_prob: Percent of adults who engage in no leisure-time physical activity</p>
</section>
<section id="classification" class="level3">
<h3 class="anchored" data-anchor-id="classification">Classification</h3>
<p>In order to support the reports hypothesis, the report asks another question:<br>
<em>Can I predict whether there are mutations in any of the obesity related genes, if a person has a cancer diagnosis based on the following data ‘age_at_diagnosis_days’, ‘primary_site’, ‘disease_type’, ‘obesity_prob’, ‘overweight_prob’, ‘nophyac_prob’, ‘musclestr2ormore_prob’?</em></p>
<p>A successful classification would help to underline the reports hypothesis which relates BRFSS and cancer data and helps to emphasize the importance on further detailed analysis.</p>
<p>Therefore the Logistic Regression is used in order to classify on a binary outcome telling whether there is a mutation in any of the genes related to obesity or not.</p>
</section>
</section>
<section id="results" class="level2">
<h2 class="anchored" data-anchor-id="results">Results</h2>
<section id="correlation-1" class="level3">
<h3 class="anchored" data-anchor-id="correlation-1">Correlation</h3>
<p>The BRFSS data in years 2011 to 2022:</p>
<div class="cell" data-execution_count="82">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>dbEngine<span class="op">=</span>sqlalchemy.create_engine(<span class="st">'sqlite:///../data/brfss.sqlite'</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> dbEngine.<span class="ex">connect</span>() <span class="im">as</span> conn:</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    sqlstatement_obese <span class="op">=</span> <span class="st">"select YearStart, Data_Value, </span><span class="ch">\"</span><span class="st">Age(years)</span><span class="ch">\"</span><span class="st">  from brfss where LocationDesc = 'National' and Total = 'Total' and Question like '%aged 18 years and older who have obesity' order by YearStart"</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    brfss_obesity2 <span class="op">=</span> pd.read_sql_query(sqlstatement_obese, conn)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    brfss_obesity2[<span class="st">'Data_Value'</span>] <span class="op">=</span> brfss_obesity2[<span class="st">'Data_Value'</span>].astype(<span class="st">"float"</span>)</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    brfss_obesity2[<span class="st">'Categories of BRFSS'</span>] <span class="op">=</span> <span class="st">'Obesity'</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>    sqlstatement_overweight <span class="op">=</span> <span class="st">"select YearStart, Data_Value, </span><span class="ch">\"</span><span class="st">Age(years)</span><span class="ch">\"</span><span class="st">  from brfss where LocationDesc = 'National' and Total = 'Total' and Question like 'Percent of adults aged 18 years and older who have an overweight classification' order by YearStart"</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>    brfss_overweight <span class="op">=</span> pd.read_sql_query(sqlstatement_overweight, conn)</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>    brfss_overweight[<span class="st">'Data_Value'</span>] <span class="op">=</span> brfss_overweight[<span class="st">'Data_Value'</span>].astype(<span class="st">"float"</span>)</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>    brfss_overweight[<span class="st">'Categories of BRFSS'</span>] <span class="op">=</span> <span class="st">'Overweight'</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>    sqlstatement_ms2d <span class="op">=</span> <span class="st">"select YearStart, Data_Value, </span><span class="ch">\"</span><span class="st">Age(years)</span><span class="ch">\"</span><span class="st">  from brfss where LocationDesc = 'National' and Total = 'Total' and Question like 'Percent of adults who engage in muscle-strengthening activities on 2 or more days a week' order by YearStart"</span></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>    brfss_ms2d <span class="op">=</span> pd.read_sql_query(sqlstatement_ms2d, conn)</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>    brfss_ms2d[<span class="st">'Data_Value'</span>] <span class="op">=</span> brfss_ms2d[<span class="st">'Data_Value'</span>].astype(<span class="st">"float"</span>)</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>    brfss_ms2d[<span class="st">'Categories of BRFSS'</span>] <span class="op">=</span> <span class="st">'Muscle-strengthening &gt;=2 days/week'</span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>    sqlstatement_nophyac <span class="op">=</span> <span class="st">"select YearStart, Data_Value, </span><span class="ch">\"</span><span class="st">Age(years)</span><span class="ch">\"</span><span class="st">  from brfss where LocationDesc = 'National' and Total = 'Total' and Question like 'Percent of adults who engage in no leisure-time physical activity' order by YearStart"</span></span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>    brfss_nophyac <span class="op">=</span> pd.read_sql_query(sqlstatement_nophyac, conn)</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>    brfss_nophyac[<span class="st">'Data_Value'</span>] <span class="op">=</span> brfss_nophyac[<span class="st">'Data_Value'</span>].astype(<span class="st">"float"</span>)</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>    brfss_nophyac[<span class="st">'Categories of BRFSS'</span>] <span class="op">=</span> <span class="st">'No leisure-time physical activity'</span></span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a>    brfss_cat <span class="op">=</span> pd.concat([brfss_obesity2, brfss_overweight, brfss_ms2d, brfss_nophyac])</span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a>brfss_plot <span class="op">=</span> sns.catplot(</span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a>    data<span class="op">=</span>brfss_cat, x<span class="op">=</span><span class="st">"YearStart"</span>, y<span class="op">=</span><span class="st">"Data_Value"</span>, hue<span class="op">=</span><span class="st">"Categories of BRFSS"</span>, kind<span class="op">=</span><span class="st">"bar"</span>,</span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a>    palette<span class="op">=</span><span class="st">"pastel"</span>, edgecolor<span class="op">=</span><span class="st">".6"</span>, legend<span class="op">=</span><span class="va">False</span>, height<span class="op">=</span><span class="dv">5</span>, aspect<span class="op">=</span><span class="fl">1.5</span></span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"Year"</span>)</span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Percentage"</span>)</span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a>plt.legend(loc<span class="op">=</span><span class="st">'lower right'</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="report_files/figure-html/cell-4-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>GDC data and its increase in entries of genes with mutations. Genes as referred to the selected genes related to obesity, see <a href="#genes">Data pipeline-Transformation-Genes</a>.</p>
<div class="cell" data-execution_count="83">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>dbEngine<span class="op">=</span>sqlalchemy.create_engine(<span class="st">'sqlite:///../data/gdc.sqlite'</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> dbEngine.<span class="ex">connect</span>() <span class="im">as</span> conn:</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    sqlstatement_obese <span class="op">=</span> <span class="st">"select year_of_diagnosis, count(*) as nrdiag, sum(sim), sum(p), sum(lpr), sum(mr), sum(a), sum(n), sum(m), sum(k), sum(l), sum(pc), sum(b), sum(s), sum(sim)+sum(p)+sum(lpr)+sum(mr)+sum(a)+sum(n)+sum(m)+sum(k)+sum(l)+sum(pc)+sum(b)+sum(s) as gene_sum, (sum(sim)+sum(p)+sum(lpr)+sum(mr)+sum(a)+sum(n)+sum(m)+sum(k)+sum(l)+sum(pc)+sum(b)+sum(s))/count(*) as gene_sum_per_nrdiag from (select d.year_of_diagnosis, c.</span><span class="ch">\"</span><span class="st">gene.SIM1</span><span class="ch">\"</span><span class="st"> as sim, c.</span><span class="ch">\"</span><span class="st">gene.POMC</span><span class="ch">\"</span><span class="st"> as p, c.</span><span class="ch">\"</span><span class="st">gene.LEPR</span><span class="ch">\"</span><span class="st"> as lpr, c.</span><span class="ch">\"</span><span class="st">gene.MRAP2</span><span class="ch">\"</span><span class="st"> as mr, c.</span><span class="ch">\"</span><span class="st">gene.ADCY3</span><span class="ch">\"</span><span class="st"> as a, c.</span><span class="ch">\"</span><span class="st">gene.NTRK2</span><span class="ch">\"</span><span class="st"> as n, c.</span><span class="ch">\"</span><span class="st">gene.MC4R</span><span class="ch">\"</span><span class="st"> as m, c.</span><span class="ch">\"</span><span class="st">gene.KSR2</span><span class="ch">\"</span><span class="st"> as k, c.</span><span class="ch">\"</span><span class="st">gene.LEP</span><span class="ch">\"</span><span class="st"> as l, c.</span><span class="ch">\"</span><span class="st">gene.PCSK1</span><span class="ch">\"</span><span class="st"> as pc, c.</span><span class="ch">\"</span><span class="st">gene.BDNF</span><span class="ch">\"</span><span class="st"> as b, c.</span><span class="ch">\"</span><span class="st">gene.SH2B1</span><span class="ch">\"</span><span class="st"> as s from cases c, diagnoses d where c.id = d.case_id and year_of_diagnosis is not null) group by year_of_diagnosis ;"</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    gdc <span class="op">=</span> pd.read_sql_query(sqlstatement_obese, conn)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>    gdc[<span class="st">'year_of_diagnosis'</span>] <span class="op">=</span> gdc[<span class="st">'year_of_diagnosis'</span>].astype(<span class="bu">int</span>)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>    gdc_woGene <span class="op">=</span> pd.read_sql_query(<span class="st">'select year_of_diagnosis, count(*) as nrdiag from (select d.year_of_diagnosis, c."gene.SIM1"+c."gene.POMC"+c."gene.LEPR"+c."gene.MRAP2"+c."gene.ADCY3"+c."gene.NTRK2"+c."gene.MC4R"+c."gene.KSR2"+c."gene.LEP"+c."gene.PCSK1"+c."gene.BDNF"+c."gene.SH2B1" as gene_sum from cases c, diagnoses d where c.id = d.case_id and year_of_diagnosis is not null) where gene_sum = 0 group by year_of_diagnosis;'</span>, conn)</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>    gdc_woGene[<span class="st">'year_of_diagnosis'</span>] <span class="op">=</span> gdc[<span class="st">'year_of_diagnosis'</span>].astype(<span class="bu">int</span>)</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>ax2 <span class="op">=</span> ax.twinx()</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'year_of_diagnosis'</span>, y<span class="op">=</span><span class="st">'nrdiag'</span>, data<span class="op">=</span>gdc, ax<span class="op">=</span>ax2, label<span class="op">=</span><span class="st">'Number of diagnoses'</span>, legend<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'year_of_diagnosis'</span>, y<span class="op">=</span><span class="st">'nrdiag'</span>, data<span class="op">=</span>gdc_woGene, ax<span class="op">=</span>ax2, color<span class="op">=</span><span class="st">'green'</span>, label<span class="op">=</span><span class="st">'Number of diagnoses without mentioning specific genes'</span>, legend<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>legax2 <span class="op">=</span> plt.legend()</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span><span class="st">'year_of_diagnosis'</span>, y<span class="op">=</span><span class="st">'gene_sum'</span>, data<span class="op">=</span>gdc, ax<span class="op">=</span>ax, color<span class="op">=</span><span class="st">'r'</span>, label<span class="op">=</span><span class="st">'Summation of mutations in specific genes'</span>, legend<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>legax <span class="op">=</span> plt.legend()</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>legax2.remove()</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>legax.remove()</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>ax2.set_ylim(<span class="dv">0</span>, <span class="dv">2000</span>)</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>ax.set_ylim(<span class="dv">0</span>, <span class="dv">7000</span>)</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>ax.<span class="bu">set</span>(xlabel<span class="op">=</span><span class="st">'Year of diagnosis'</span>, ylabel <span class="op">=</span> <span class="st">'Sum of mutations over all diagnoses'</span>)</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>ax2.<span class="bu">set</span>(ylabel<span class="op">=</span><span class="st">'Number of diagnoses'</span>)</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>plt.figlegend(loc<span class="op">=</span><span class="st">'upper left'</span>, fancybox<span class="op">=</span><span class="va">True</span>, framealpha<span class="op">=</span><span class="dv">1</span>, shadow<span class="op">=</span><span class="va">True</span>, borderpad<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="report_files/figure-html/cell-5-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>In the graph there is a relation of mutation in genes (red) to the number of diagnoses (blue), however, there are diagnoses without mentioning any gene mutations at all (green) which are included in the overal number of diagnoses (blue). For many other assumptions in this report here also applies that the submissions to GDC are on a free basis and that means there is no statement about the completeness of the data whether the diagnoses are incomplete (missing genes) or there are no mutations.</p>
<div class="cell" data-execution_count="84">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>dbEngine<span class="op">=</span>sqlalchemy.create_engine(<span class="st">'sqlite:///../data/gdc.sqlite'</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> dbEngine.<span class="ex">connect</span>() <span class="im">as</span> conn:</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>    sqlstatement_obese <span class="op">=</span> <span class="st">"select d.*, c.*,</span><span class="ch">\</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="st">  case</span><span class="ch">\</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="st">   when age_at_diagnosis_years &lt;18 then 'Under 18'</span><span class="ch">\</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="st">   when age_at_diagnosis_years between 18 and 24 then '18 - 24'</span><span class="ch">\</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="st">   when age_at_diagnosis_years between 25 and 34 then '25 - 34'</span><span class="ch">\</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="st">   when age_at_diagnosis_years between 35 and 44 then '35 - 44'</span><span class="ch">\</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="st">   when age_at_diagnosis_years between 45 and 54 then '45 - 54'</span><span class="ch">\</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a><span class="st">   when age_at_diagnosis_years between 55 and 64 then '55 - 64'</span><span class="ch">\</span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a><span class="st">   when age_at_diagnosis_years &gt;= 65 then '65 or older'</span><span class="ch">\</span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a><span class="st"> END as age_range,</span><span class="ch">\</span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a><span class="st"> CASE </span><span class="ch">\</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a><span class="st">    when c.</span><span class="ch">\"</span><span class="st">gene.SIM1</span><span class="ch">\"</span><span class="st"> = '1' or c.</span><span class="ch">\"</span><span class="st">gene.POMC</span><span class="ch">\"</span><span class="st"> = 1 or c.</span><span class="ch">\"</span><span class="st">gene.LEPR</span><span class="ch">\"</span><span class="st"> ='1' or c.</span><span class="ch">\"</span><span class="st">gene.MRAP2</span><span class="ch">\"</span><span class="st"> = 1 or c.</span><span class="ch">\"</span><span class="st">gene.ADCY3</span><span class="ch">\"</span><span class="st"> = 1 or c.</span><span class="ch">\"</span><span class="st">gene.NTRK2</span><span class="ch">\"</span><span class="st"> = 1 or c.</span><span class="ch">\"</span><span class="st">gene.MC4R</span><span class="ch">\"</span><span class="st">  = 1 or c.</span><span class="ch">\"</span><span class="st">gene.KSR2</span><span class="ch">\"</span><span class="st"> = 1 or c.</span><span class="ch">\"</span><span class="st">gene.LEP</span><span class="ch">\"</span><span class="st"> = 1 or c.</span><span class="ch">\"</span><span class="st">gene.PCSK1</span><span class="ch">\"</span><span class="st"> = 1 or c.</span><span class="ch">\"</span><span class="st">gene.BDNF</span><span class="ch">\"</span><span class="st">  =1 or c.</span><span class="ch">\"</span><span class="st">gene.SH2B1</span><span class="ch">\"</span><span class="st"> = 1 then TRUE </span><span class="ch">\</span></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a><span class="st">    ELSE FALSE </span><span class="ch">\</span></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a><span class="st"> END as obesity_mutation,</span><span class="ch">\</span></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a><span class="st"> (c.</span><span class="ch">\"</span><span class="st">gene.SIM1</span><span class="ch">\"</span><span class="st"> + c.</span><span class="ch">\"</span><span class="st">gene.POMC</span><span class="ch">\"</span><span class="st"> + c.</span><span class="ch">\"</span><span class="st">gene.LEPR</span><span class="ch">\"</span><span class="st"> + c.</span><span class="ch">\"</span><span class="st">gene.MRAP2</span><span class="ch">\"</span><span class="st"> + c.</span><span class="ch">\"</span><span class="st">gene.ADCY3</span><span class="ch">\"</span><span class="st"> + c.</span><span class="ch">\"</span><span class="st">gene.NTRK2</span><span class="ch">\"</span><span class="st"> + c.</span><span class="ch">\"</span><span class="st">gene.MC4R</span><span class="ch">\"</span><span class="st"> + c.</span><span class="ch">\"</span><span class="st">gene.KSR2</span><span class="ch">\"</span><span class="st"> + c.</span><span class="ch">\"</span><span class="st">gene.LEP</span><span class="ch">\"</span><span class="st"> + c.</span><span class="ch">\"</span><span class="st">gene.PCSK1</span><span class="ch">\"</span><span class="st"> + c.</span><span class="ch">\"</span><span class="st">gene.BDNF</span><span class="ch">\"</span><span class="st"> + c.</span><span class="ch">\"</span><span class="st">gene.SH2B1</span><span class="ch">\"</span><span class="st">) as obesity_mutation_count</span><span class="ch">\</span></span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a><span class="st"> from diagnoses d, cases c</span><span class="ch">\</span></span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a><span class="st"> where d.case_id = c.id  and year_of_diagnosis = '2011' and year_of_diagnosis is not null;"</span></span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>    gdc <span class="op">=</span> pd.read_sql_query(sqlstatement_obese, conn)</span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>    gdc[<span class="st">'year_of_diagnosis'</span>] <span class="op">=</span> gdc[<span class="st">'year_of_diagnosis'</span>].astype(<span class="bu">int</span>)</span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a>dbEngine<span class="op">=</span>sqlalchemy.create_engine(<span class="st">'sqlite:///../data/brfss.sqlite'</span>)</span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> dbEngine.<span class="ex">connect</span>() <span class="im">as</span> conn:</span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a>    sqlstatement_musclestr2ormore <span class="op">=</span> <span class="st">"select YearStart, Data_Value, </span><span class="ch">\"</span><span class="st">Age(years)</span><span class="ch">\"</span><span class="st">  from brfss where LocationDesc = 'National' and length(</span><span class="ch">\"</span><span class="st">Age(years)</span><span class="ch">\"</span><span class="st">) &gt; 0 and Question like 'Percent of adults who engage in muscle-strengthening activities on 2 or more days a week' and YearStart = 2011 order by StratificationID1;"</span></span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a>    brfss_musclestr2ormore <span class="op">=</span> pd.read_sql_query(sqlstatement_musclestr2ormore, conn)</span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true" tabindex="-1"></a>    brfss_musclestr2ormore[<span class="st">'Data_Value'</span>] <span class="op">=</span> brfss_musclestr2ormore[<span class="st">'Data_Value'</span>].astype(<span class="st">"float"</span>)</span>
<span id="cb6-28"><a href="#cb6-28" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb6-29"><a href="#cb6-29" aria-hidden="true" tabindex="-1"></a>    sqlstatement_nophyac <span class="op">=</span> <span class="st">"select YearStart, Data_Value, </span><span class="ch">\"</span><span class="st">Age(years)</span><span class="ch">\"</span><span class="st">  from brfss where LocationDesc = 'National' and length(</span><span class="ch">\"</span><span class="st">Age(years)</span><span class="ch">\"</span><span class="st">) &gt; 0 and Question like 'Percent of adults who engage in no leisure-time physical activity' and YearStart = 2011 order by StratificationID1;"</span></span>
<span id="cb6-30"><a href="#cb6-30" aria-hidden="true" tabindex="-1"></a>    brfss_nophyac <span class="op">=</span> pd.read_sql_query(sqlstatement_nophyac, conn)</span>
<span id="cb6-31"><a href="#cb6-31" aria-hidden="true" tabindex="-1"></a>    brfss_nophyac[<span class="st">'Data_Value'</span>] <span class="op">=</span> brfss_nophyac[<span class="st">'Data_Value'</span>].astype(<span class="st">"float"</span>)</span>
<span id="cb6-32"><a href="#cb6-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-33"><a href="#cb6-33" aria-hidden="true" tabindex="-1"></a>    sqlstatement_overweight <span class="op">=</span> <span class="st">"select YearStart, Data_Value, </span><span class="ch">\"</span><span class="st">Age(years)</span><span class="ch">\"</span><span class="st">  from brfss where LocationDesc = 'National' and length(</span><span class="ch">\"</span><span class="st">Age(years)</span><span class="ch">\"</span><span class="st">) &gt; 0 and Question like 'Percent of adults aged 18 years and older who have an overweight classification' and YearStart = 2011 order by StratificationID1;"</span></span>
<span id="cb6-34"><a href="#cb6-34" aria-hidden="true" tabindex="-1"></a>    brfss_overweight <span class="op">=</span> pd.read_sql_query(sqlstatement_overweight, conn)</span>
<span id="cb6-35"><a href="#cb6-35" aria-hidden="true" tabindex="-1"></a>    brfss_overweight[<span class="st">'Data_Value'</span>] <span class="op">=</span> brfss_overweight[<span class="st">'Data_Value'</span>].astype(<span class="st">"float"</span>)</span>
<span id="cb6-36"><a href="#cb6-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-37"><a href="#cb6-37" aria-hidden="true" tabindex="-1"></a>    sqlstatement_obese <span class="op">=</span> <span class="st">"select YearStart, Data_Value, </span><span class="ch">\"</span><span class="st">Age(years)</span><span class="ch">\"</span><span class="st">  from brfss where LocationDesc = 'National' and length(</span><span class="ch">\"</span><span class="st">Age(years)</span><span class="ch">\"</span><span class="st">) &gt; 0 and Question like '%aged 18 years and older who have obesity' and YearStart = 2011 order by StratificationID1;"</span></span>
<span id="cb6-38"><a href="#cb6-38" aria-hidden="true" tabindex="-1"></a>    brfss_obesity <span class="op">=</span> pd.read_sql_query(sqlstatement_obese, conn)</span>
<span id="cb6-39"><a href="#cb6-39" aria-hidden="true" tabindex="-1"></a>    brfss_obesity[<span class="st">'Data_Value'</span>] <span class="op">=</span> brfss_obesity[<span class="st">'Data_Value'</span>].astype(<span class="st">"float"</span>)</span>
<span id="cb6-40"><a href="#cb6-40" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb6-41"><a href="#cb6-41" aria-hidden="true" tabindex="-1"></a>brfss_gdc <span class="op">=</span> gdc</span>
<span id="cb6-42"><a href="#cb6-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-43"><a href="#cb6-43" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> obesity_prob(age_group):</span>
<span id="cb6-44"><a href="#cb6-44" aria-hidden="true" tabindex="-1"></a>    ob_lines <span class="op">=</span> brfss_obesity.loc[brfss_obesity[<span class="st">'Age(years)'</span>] <span class="op">==</span> age_group]</span>
<span id="cb6-45"><a href="#cb6-45" aria-hidden="true" tabindex="-1"></a>    ob_line <span class="op">=</span> ob_lines.iloc[<span class="dv">0</span>]</span>
<span id="cb6-46"><a href="#cb6-46" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> ob_line[<span class="st">'Data_Value'</span>]</span>
<span id="cb6-47"><a href="#cb6-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-48"><a href="#cb6-48" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> overweight_prob(age_group):</span>
<span id="cb6-49"><a href="#cb6-49" aria-hidden="true" tabindex="-1"></a>    lines <span class="op">=</span> brfss_overweight.loc[brfss_overweight[<span class="st">'Age(years)'</span>] <span class="op">==</span> age_group]</span>
<span id="cb6-50"><a href="#cb6-50" aria-hidden="true" tabindex="-1"></a>    line <span class="op">=</span> lines.iloc[<span class="dv">0</span>]</span>
<span id="cb6-51"><a href="#cb6-51" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> line[<span class="st">'Data_Value'</span>]</span>
<span id="cb6-52"><a href="#cb6-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-53"><a href="#cb6-53" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> nophyac_prob(age_group):</span>
<span id="cb6-54"><a href="#cb6-54" aria-hidden="true" tabindex="-1"></a>    lines <span class="op">=</span> brfss_nophyac.loc[brfss_nophyac[<span class="st">'Age(years)'</span>] <span class="op">==</span> age_group]</span>
<span id="cb6-55"><a href="#cb6-55" aria-hidden="true" tabindex="-1"></a>    line <span class="op">=</span> lines.iloc[<span class="dv">0</span>]</span>
<span id="cb6-56"><a href="#cb6-56" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> line[<span class="st">'Data_Value'</span>]</span>
<span id="cb6-57"><a href="#cb6-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-58"><a href="#cb6-58" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> musclestr2ormore_prob(age_group):</span>
<span id="cb6-59"><a href="#cb6-59" aria-hidden="true" tabindex="-1"></a>    lines <span class="op">=</span> brfss_musclestr2ormore.loc[brfss_musclestr2ormore[<span class="st">'Age(years)'</span>] <span class="op">==</span> age_group]</span>
<span id="cb6-60"><a href="#cb6-60" aria-hidden="true" tabindex="-1"></a>    line <span class="op">=</span> lines.iloc[<span class="dv">0</span>]</span>
<span id="cb6-61"><a href="#cb6-61" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> line[<span class="st">'Data_Value'</span>]</span>
<span id="cb6-62"><a href="#cb6-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-63"><a href="#cb6-63" aria-hidden="true" tabindex="-1"></a>brfss_gdc[<span class="st">'obesity_prob'</span>] <span class="op">=</span> brfss_gdc[<span class="st">'age_range'</span>].<span class="bu">map</span>(obesity_prob)</span>
<span id="cb6-64"><a href="#cb6-64" aria-hidden="true" tabindex="-1"></a>brfss_gdc[<span class="st">'overweight_prob'</span>] <span class="op">=</span> brfss_gdc[<span class="st">'age_range'</span>].<span class="bu">map</span>(overweight_prob)</span>
<span id="cb6-65"><a href="#cb6-65" aria-hidden="true" tabindex="-1"></a>brfss_gdc[<span class="st">'nophyac_prob'</span>] <span class="op">=</span> brfss_gdc[<span class="st">'age_range'</span>].<span class="bu">map</span>(nophyac_prob)</span>
<span id="cb6-66"><a href="#cb6-66" aria-hidden="true" tabindex="-1"></a>brfss_gdc[<span class="st">'musclestr2ormore_prob'</span>] <span class="op">=</span> brfss_gdc[<span class="st">'age_range'</span>].<span class="bu">map</span>(musclestr2ormore_prob)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next graph shows the amount of changes in the specific genes in each age group. Since BRFSS data contains age groups, we relate the two datasets with those groups.</p>
<div class="cell" data-execution_count="85">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>f<span class="op">=</span>sns.catplot(data<span class="op">=</span>brfss_gdc, x<span class="op">=</span><span class="st">'obesity_mutation_count'</span>, y<span class="op">=</span><span class="st">'age_range'</span>, kind<span class="op">=</span><span class="st">'violin'</span>)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"Mutations"</span>)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Age group"</span>)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>f.<span class="bu">set</span>(xlim<span class="op">=</span>(<span class="dv">0</span>,<span class="dv">12</span>))<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="report_files/figure-html/cell-7-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>The graph shows that as older the patients are the more the mutations occur in the genes related to obesity.</p>
<p>Next, we have a look at the correlation matrix for the following attributes:<br>
- age_at_diagnosis_days: Age at diagnosis (days)<br>
- obesity_prob: Probability of obesity<br>
- overweight_prob: Probability of overweight<br>
- nophyac_prob: Probability of no physical activity<br>
- musclestr2ormore_prob: Probability of muscle strengthening 2 or more days a week<br>
- obesity_mutation: Mutations in genes related to obesity (y/n)<br>
- obesity_mutation_count: Number of mutations in genes related to obesity</p>
<div class="cell" data-execution_count="86">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>xdf <span class="op">=</span> pd.DataFrame(brfss_gdc, columns<span class="op">=</span>[<span class="st">'age_at_diagnosis_days'</span>, <span class="st">'obesity_prob'</span>, <span class="st">'overweight_prob'</span>, <span class="st">'nophyac_prob'</span>, <span class="st">'musclestr2ormore_prob'</span>, <span class="st">'obesity_mutation'</span>, <span class="st">'obesity_mutation_count'</span>])</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>correlation_df <span class="op">=</span> xdf</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>correlation_matrix <span class="op">=</span> correlation_df.corr()</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">8</span>))</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>sns.heatmap(correlation_matrix, annot<span class="op">=</span><span class="va">True</span>, cmap<span class="op">=</span><span class="st">'coolwarm'</span>, fmt<span class="op">=</span><span class="st">".2f"</span>, linewidths<span class="op">=</span><span class="fl">.5</span>)</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">'Correlation Heatmap: Age at Diagnosis/Mutations in Genes with Probabilities of Behavioural Risks'</span>)</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="report_files/figure-html/cell-8-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>The correlation matrix shows correlation between:<br>
Age at diagnosis and<br>
- Probability of overweight (0.83)<br>
- Probability of no physical activity (0.94)<br>
- Probability of muscle strengthening 2 or more days a week (-0.90)<br>
Other combination do not show significant correlation.</p>
</section>
<section id="classification-1" class="level3">
<h3 class="anchored" data-anchor-id="classification-1">Classification</h3>
<p>For the classification, the logistic regression is used to classify if there is a mutation in one or more genes related to obesity or not. So, it is a binary classification and the logistic regression is a perfect fit.</p>
<div class="cell" data-execution_count="87">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>xdf <span class="op">=</span> pd.DataFrame(brfss_gdc, columns<span class="op">=</span>[<span class="st">'age_at_diagnosis_days'</span>, <span class="st">'primary_site'</span>, <span class="st">'disease_type'</span>, <span class="st">'obesity_prob'</span>, <span class="st">'overweight_prob'</span>, <span class="st">'nophyac_prob'</span>, <span class="st">'musclestr2ormore_prob'</span>])</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> onehotEncoding(orig, feature):</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>    one_hot <span class="op">=</span> pd.get_dummies(xdf[[feature]])</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>    res <span class="op">=</span> pd.concat([xdf, one_hot], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span>(res)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>xdf <span class="op">=</span> onehotEncoding(xdf, <span class="st">'primary_site'</span>)</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>xdf <span class="op">=</span> onehotEncoding(xdf, <span class="st">'disease_type'</span>)</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>xdf.pop(<span class="st">'primary_site'</span>)</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>xdf.pop(<span class="st">'disease_type'</span>)<span class="op">;</span></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>y <span class="op">=</span> brfss_gdc[<span class="st">'obesity_mutation'</span>]</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>x_train, x_temp, y_train, y_temp <span class="op">=</span> train_test_split(xdf, y, random_state<span class="op">=</span><span class="dv">42</span>, train_size<span class="op">=</span><span class="fl">.7</span>, test_size<span class="op">=</span><span class="fl">.3</span>)</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>x_validation, x_test, y_validation, y_test <span class="op">=</span> train_test_split(x_temp, y_temp, random_state<span class="op">=</span><span class="dv">42</span>, train_size<span class="op">=</span><span class="dv">2</span><span class="op">/</span><span class="dv">3</span>, test_size<span class="op">=</span><span class="dv">1</span><span class="op">/</span><span class="dv">3</span>)</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a><span class="co">#Logistic regression</span></span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>logreg <span class="op">=</span> LogisticRegression(max_iter <span class="op">=</span> <span class="dv">10000</span>)</span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>logreg.fit(x_train, y_train)<span class="op">;</span></span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a><span class="co">#Validation</span></span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a>prediction <span class="op">=</span> logreg.predict(x_validation)</span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a><span class="co">##Classification Report</span></span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(classification_report(y_validation, prediction))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              precision    recall  f1-score   support

           0       0.90      0.47      0.62        81
           1       0.85      0.98      0.91       249

    accuracy                           0.86       330
   macro avg       0.88      0.73      0.77       330
weighted avg       0.86      0.86      0.84       330
</code></pre>
</div>
</div>
<p>The validation of the logistic regression shows the precision, recall and f1-score of the classification for<br>
- 0: no mutation<br>
- 1: mutation in one or more genes related to obesity<br>
The sampling size is 81 for class=0 and 249 for class=1, so a validation set of size 330. We splitted the data into a training set (1155 diagnoses), a validation set (330 diagnoses) and a test set(166 diagnoses).</p>
<p>The training set is used to train the logistic regression function, the validation set is used to validate the predictions and the test set is kept away from the training phase in order to prevent overfitting.<br>
The test set would come into play when a model selection would occur. Since there is only one model in the report the test set is not used.</p>
<p>Further validation of the logistic regression is done by providing the Receiver-Operator-Curve and the Area-Under-Curve value:</p>
<div class="cell" data-execution_count="88">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co">##ROC and AUC</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>y_pred_proba <span class="op">=</span> logreg.predict_proba(x_validation)[::,<span class="dv">1</span>]</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>fpr, tpr, _ <span class="op">=</span> roc_curve(y_validation,  y_pred_proba)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>auc <span class="op">=</span> roc_auc_score(y_validation, y_pred_proba)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="co">#create ROC curve</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>plt.plot(fpr,tpr,label<span class="op">=</span><span class="st">"AUC="</span><span class="op">+</span><span class="bu">str</span>(auc))</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">'True Positive Rate'</span>)</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">'False Positive Rate'</span>)</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>plt.legend(loc<span class="op">=</span><span class="dv">4</span>)</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="report_files/figure-html/cell-10-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
</section>
<section id="discussion-and-conclusions" class="level2">
<h2 class="anchored" data-anchor-id="discussion-and-conclusions">Discussion and Conclusions</h2>
<p>Limitations on the GDC data caused that only the year 2011 is included into the report due to no completeness on cancer data delivered to GDC and not enough years overlapping in both datasets.</p>
<p>Even though the GDC Datasource has a field for exposure weight or BMI, it is not filled in any of their datasets. Therefore, it is not possible to link the obesity numbers of BRFSS dataset to the cancer in general.</p>
<p>The BRFSS data is gathered from citizens in the United States. In contrast to the GDC dataset where diagnoses are linked to different countries (united states, canada, …). Since a lot of data sets in GDC are without any specification on that field (country of residence at enrollment), for reasons of ease for this project the cases from GDC are not restricted to United States.</p>
<p>With these limitations, the report can show that there may be correlation between behavioural risk factors (overweight, no physical activity and muscle strengthening 2 or more days a week) and the age at the diagnosis of the cancer.</p>
<p>The correlation matrix shows that there can be a classification task which could potentially be successful.</p>
<p>The logistic regression is trained and reveals good validation metrics.</p>
<p>The classification report for the logistic regression indicates a good precision for both classes which means that the classification can identify positive predictions on both classes quite well (the number of wrong classified cases in the classes is low - small number in False Positives).</p>
<p>The recall and then also the F1-score drops for classifying no mutations (class=0) which means that the classification is not very sensitive towards classifying diagnoses without any mutations. Hence, there will be more False Negatives for the class 0 which means there will be an amount of diagnoses which are classified as class 1 but are class 0 (some False Negatives for class=0).</p>
<p>This means that we tend to classify more towards class 1 which would be ok for further investigation since it will just include more cases to investigate the mutations in genes.</p>
<p>The Receiver-Operator-Curve ROC shows that we have a result which is ok. It is definitly better than random (diagonal). The Area-Under-Curve is around 0.83 which is also close to 1. The closer it is to 1 the better the specificity and sensitivity of the classifier.</p>
<p>This could be used to investigate on more classifiers by compareing the validation results, deciding on a classifier and running a test with the test data (kept away from train-validation cycle).</p>
<p>In overal, all the results have to be used with care and caution since cancer is influenced by many things, not only behavioural risk factors. Thus, this report can only be a direction for further investigations on this topic.</p>
<section id="outlook" class="level4">
<h4 class="anchored" data-anchor-id="outlook">Outlook</h4>
<p>The result in this report shows that mutations in genes related to obesity for patients with a cancer diagnosis could potentially be linked to the behavioural risk factors. On a next step, it would be very interesting to connect obesity identifying data (BMI) to cancer data on a case basis.</p>
<p>Futhermore, changes in genes can be harmful but can also be without or with low consequences. Therefore a deeper evaluation on the impact of the DNA changes of the obesity related genes would be beneficial. This data is also available in GDC.</p>
<p>The submissions are not forced for GDC. In Germany there is a law for delivering cancer related data, however, it only contains a very limited subset of information which GDC is providing. I would highly recommend the possibility in the future to either add obesity identifying data (like BMI) to cancer data or to make it possible to connect datasets (like in other studies, e.g.&nbsp;NAKO - Nationale Gesundheitsstudie, which is not open data).</p>
</section>
</section>
<section id="references" class="level2">
<h2 class="anchored" data-anchor-id="references">References</h2>
</section>


<div id="quarto-appendix" class="default"><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>Genes contributing to obesity <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8459824/">The genetics of obesity: from discovery to biology</a> Loos RJF, Yeo GSH. The genetics of obesity: from discovery to biology. Nat Rev Genet. 2022 Feb;23(2):120-133. doi: 10.1038/s41576-021-00414-z. Epub 2021 Sep 23. PMID: 34556834; PMCID: PMC8459824.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>